tests/testsuite/test.rs

```diff
@@ -3042,6 +3044,82 @@ fn json_artifact_includes_test_flag() {
         ).run();
 }
 
+#[test]
+fn json_artifact_includes_executable_for_library_tests() {
+    let p = project()
+        .file("src/main.rs", "fn main() { }")
+        .file("src/lib.rs", r#"#[test] fn lib_test() {}"#)
+        .build();
+
+    p.cargo("test --lib -v --no-run --message-format=json")
+        .with_json(r#"
+            {
+                "executable": "[..]/foo/target/debug/foo-[..][EXE]",
+                "features": [],
+                "filenames": "{...}",
+                "fresh": false,
+                "package_id": "foo 0.0.1 ([..])",
+                "profile": "{...}",
+                "reason": "compiler-artifact",
+                "target": {
+                    "crate_types": [ "lib" ],
+                    "kind": [ "lib" ],
+                    "edition": "2015",
+                    "name": "foo",
+                    "src_path": "[..]/foo/src/lib.rs"
+                }
+            }
+        "#)
+        .run();
+}
+
+#[test]
+fn json_artifact_includes_executable_for_integration_tests() {
+    let p = project()
+        .file("src/main.rs", "fn main() {}")
+        .file("tests/integration_test.rs", r#"#[test] fn integration_test() {}"#)
+        .build();
+
+    // Using jobs=1 to ensure that the order of messages is consistent.
```

Comment:
There is no way to force a deterministic order.  Does this test need a `main.rs`?  If it does, maybe build it in a separate step?

Reply:
Looks like it isn't needed.